drivers/soft_uart/soft_uart.c

```diff
@@ -0,0 +1,323 @@
+        ctx->data_bits = 6;
+        break;
+    case UART_DATA_BITS_7:
+        ctx->data_bits = 7;
+        break;
+    case UART_DATA_BITS_8:
+        ctx->data_bits = 8;
+        break;
+    }
+
+    switch (parity) {
+    case UART_PARITY_NONE:
+        ctx->parity = PARITY_NONE;
+        break;
+    case UART_PARITY_EVEN:
+        ctx->parity = PARITY_EVEN;
+        break;
+    case UART_PARITY_ODD:
+        ctx->parity = PARITY_ODD;
+        break;
+    case UART_PARITY_MARK:
+        ctx->parity = PARITY_MARK;
+        break;
+    case UART_PARITY_SPACE:
+        ctx->parity = PARITY_SPACE;
+        break;
+    }
+
+    switch (stop_bits) {
+    case UART_STOP_BITS_1:
+        ctx->stop_bits = 1;
+        break;
+    case UART_STOP_BITS_2:
+        ctx->stop_bits = 2;
+        break;
+    }
+
+    return 0;
+}
+#endif /* MODULE_SOFT_UART_MODECF */
+
+static void soft_uart_write_byte(soft_uart_t uart, uint8_t data)
+{
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    /* start bit (LOW) + data bits */
+    ctx->bits_tx = 1 + BITS_DATA(ctx);
+    ctx->byte_tx = data << 1;
+
+#ifdef MODULE_SOFT_UART_MODECFG
+    if (ctx->parity != PARITY_NONE) {
+        uint8_t parity = 0;
+
+        switch (ctx->parity) {
+        case PARITY_EVEN:
+            parity = __builtin_parity(data);
+            break;
+        case PARITY_ODD:
+            parity = !__builtin_parity(data);
+            break;
+        case PARITY_MARK:
+            parity = 1;
+            break;
+        case PARITY_SPACE:
+            parity = 0;
+            break;
+        }
+
+        ctx->byte_tx |= parity << ctx->bits_tx++;
+    }
+#endif
+
+    for (int i = 0; i < BITS_STOP(ctx); ++i) {
+        ctx->byte_tx |= 1 << ctx->bits_tx++;
+    }
+
+    if (cfg->flags & SOFT_UART_FLAG_INVERT_TX) {
+        ctx->byte_tx = ~ctx->byte_tx;
+    }
+
+    timer_set_periodic(cfg->tx_timer, 0, ctx->baud,
+                       TIM_FLAG_RESET_ON_MATCH | TIM_FLAG_RESET_ON_SET);
+    mutex_lock(&ctx->sync);
+}
+
+void soft_uart_write(uart_t uart, const uint8_t *data, size_t len)
+{
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    const uint8_t *end = data + len;
+
+    mutex_lock(&ctx->lock);
+    timer_start(cfg->tx_timer);
+
+    while (data != end) {
+        soft_uart_write_byte(uart, *data++);
+    }
+
+    timer_stop(cfg->tx_timer);
+    mutex_unlock(&soft_uart_ctx[uart].lock);
+}
+
+void soft_uart_poweron(soft_uart_t uart)
+{
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    if (ctx->rx_cb) {
+        gpio_irq_enable(cfg->rx_pin);
+    }
+}
+
+void soft_uart_poweroff(soft_uart_t uart)
+{
+    const soft_uart_conf_t *cfg = &soft_uart_config[uart];
+    struct uart_ctx *ctx = &soft_uart_ctx[uart];
+
+    if (ctx->rx_cb) {
+        gpio_irq_disable(cfg->rx_pin);
+    }
+}
```

Comment:
What about the timer(s)?

Reply:
The RX timer is stopped in `_rx_timer_cb`.
The TX timer should be stopped in `soft_uart_write`

Reply:
OK. Sorry, missed that.